clients/js/src/compat.ts

```diff
@@ -505,11 +507,20 @@ async function inferRuntimePublicKeySource(
 ): Promise<Parameters<typeof fetchRuntimePublicKey>[0]> {
   const isSigner = isEthersSigner(upstream);
   if (isSigner || isEthersProvider(upstream)) {
-    return {
-      chainId: isSigner
-        ? await upstream.getChainId()
-        : (await upstream.getNetwork()).chainId,
-    };
+    const chainId = isSigner
+      ? await upstream.getChainId()
+      : (await upstream.getNetwork()).chainId;
+    // for hardhat localhost testing network, directly return its ethersProvider
+    if (
+      chainId == HARDHAT_LOCAL_NETWORK_CHAINID &&
+      'provider' in upstream &&
+      upstream['provider'] &&
+      'send' in upstream['provider']
+    ) {
+      return upstream['provider'];
+    } else {
+      return { chainId: chainId };
+    }
```

Comment:
This is a good idea, but it doesn't generalize since the hardhat network chain id can actually be set by the developer. How about this: try to get the public key using the provider, and if that fails fall back to the oasis web3 gateways.